clang/test/CIR/CodeGen/struct.c

```diff
@@ -61,3 +61,13 @@ struct S3 {
   int a;
 } s3[3] = {{1}, {2}, {3}};
 // CHECK-DAG: cir.global external @s3 = #cir.const_array<[#cir.const_struct<{#cir.int<1> : !s32i}> : !ty_22struct2ES322, #cir.const_struct<{#cir.int<2> : !s32i}> : !ty_22struct2ES322, #cir.const_struct<{#cir.int<3> : !s32i}> : !ty_22struct2ES322]> : !cir.array<!ty_22struct2ES322 x 3>
+
+struct Bar shouldGenerateAndAccessStructArrays(void) {
+  struct Bar s[1] = {{3, 4}};
```

Comment:
Do ctors show up at all in the AST for this or is the aggregate init semantics are more direct at the AST level? Ideally ctors that are transformed directly to memcpy's by Clang should live in the ctor form until some other pass in the pipeline (e.g. LLVMLoweringPrepare) makes them into cir.copy's before LLVM lowering (or other relevant analysis). One of the current problems with Clang is that it lower those too early, which makes analysis harder (at the point where all things are memcpy's, it becomes harder to track what they're really about).

Reply:
> Do ctors show up at all in the AST for this or is the aggregate init semantics are more direct at the AST level?

They do show up, but I'm not dealing with any complex copies (e.g. copy/assignment ctors). The scenario I'm handling are only basic C copies that are converted from the AST directly to a `memcpy`.

In the original codegen, complex copies appear here:

https://github.com/llvm/clangir/blob/74979f9d4a82c9d2234bfd2b687af39e6c0cc870/clang/lib/CIR/CodeGen/CIRGenExprAgg.cpp#L233-L242

Reply:
For C is fine cause no ctors. But for C++ calling simple init ctors should be enough of a requirement to track. 

What's the difference between AST for C versus C++? Can you share snippets of them?

Reply:
@bcardosolopes the AST for:

```c++
struct S { int a; };
void foo(void) {  struct S s = {3}; }
```

<details>
  <summary>C AST</summary>

```
|-RecordDecl 0x555d8a7e99f0 </tmp/test.c:1:1, col:19> col:8 struct S definition
| `-FieldDecl 0x555d8a7e9aa8 <col:12, col:16> col:16 a 'int'
`-FunctionDecl 0x555d8a7e9bb8 <line:2:1, col:37> col:6 foo 'void (void)'
  `-CompoundStmt 0x555d8a7e9e40 <col:16, col:37>
    `-DeclStmt 0x555d8a7e9e28 <col:19, col:35>
      `-VarDecl 0x555d8a7e9d00 <col:19, col:34> col:28 s 'struct S':'struct S' cinit
        `-InitListExpr 0x555d8a7e9de0 <col:32, col:34> 'struct S':'struct S'
          `-IntegerLiteral 0x555d8a7e9d68 <col:33> 'int' 3
```

</details>

<details>
  <summary>C++ AST</summary>

 ```
|-CXXRecordDecl 0x5614151a9260 </tmp/test.cpp:1:1, col:19> col:8 struct S definition
| |-DefinitionData pass_in_registers aggregate standard_layout trivially_copyable pod trivial literal
| | |-DefaultConstructor exists trivial needs_implicit
| | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
| | |-MoveConstructor exists simple trivial needs_implicit
| | |-CopyAssignment simple trivial has_const_param needs_implicit implicit_has_const_param
| | |-MoveAssignment exists simple trivial needs_implicit
| | `-Destructor simple irrelevant trivial needs_implicit
| |-CXXRecordDecl 0x5614151a9378 <col:1, col:8> col:8 implicit struct S
| `-FieldDecl 0x5614151a9420 <col:12, col:16> col:16 a 'int'
`-FunctionDecl 0x5614151a9548 <line:2:1, col:37> col:6 foo 'void ()'
  `-CompoundStmt 0x5614151a97c0 <col:16, col:37>
    `-DeclStmt 0x5614151a97a8 <col:19, col:35>
      `-VarDecl 0x5614151a9690 <col:19, col:34> col:28 s 'struct S':'S' cinit
        `-InitListExpr 0x5614151a9760 <col:32, col:34> 'struct S':'S'
          `-IntegerLiteral 0x5614151a96f8 <col:33> 'int' 3
```
  
</details>

In C++, the struct does have trivial constructors, but these are never called in the function. But if we add an assignment:

```c++
struct S { int a; };
void foo(void) {  struct S s = {3}; }
```

<details>
  <summary>C++ AST w/ Assignment</summary>

```
`-FunctionDecl 0x560ea8d5e548 <line:2:1, line:5:1> line:2:6 foo 'void ()'
  `-CompoundStmt 0x560ea8d5f018 <col:16, line:5:1>
    |-DeclStmt 0x560ea8d5e7a8 <line:3:3, col:19>
    | `-VarDecl 0x560ea8d5e690 <col:3, col:18> col:12 used s 'struct S':'S' cinit
    |   `-InitListExpr 0x560ea8d5e760 <col:16, col:18> 'struct S':'S'
    |     `-IntegerLiteral 0x560ea8d5e6f8 <col:17> 'int' 3
    `-DeclStmt 0x560ea8d5f000 <line:4:3, col:17>
      `-VarDecl 0x560ea8d5e7e0 <col:3, col:16> col:12 a 'struct S':'S' cinit
        `-CXXConstructExpr 0x560ea8d5efd0 <col:16> 'struct S':'S' 'void (const S &) noexcept'
          `-ImplicitCastExpr 0x560ea8d5ed88 <col:16> 'const S':'const S' lvalue <NoOp>
            `-DeclRefExpr 0x560ea8d5e848 <col:16> 'struct S':'S' lvalue Var 0x560ea8d5e690 's' 'struct S':'S'
```

</details>

Then we see a `CXXConstructExpr` node in the function.

Reply:
Thanks for the detailed explanation.

> But if we add an assignment:

The code snippet looks the same to me, what am I missing?

Reply:
@bcardosolopes sorry, I forgot to update the code. It's fixed now.

Reply:
Np. Ok, we are good for non-assignment! For the assignment one, do we crash or emit appropriate ctor call? If it crashes, no need to handle as part of this patch, I just wanna make sure we're not short-circuiting the assignment in case it ctor is explicitly used in the AST nodes.